### PR TITLE
[Semi-Modular] Adds back TEG 

### PR DESCRIPTION
**Pull request by Erol509**
## About The Pull Request

Ports TEG from Ship Test ( Code) / Yog station (visuals). 

https://github.com/shiptest-ss13/Shiptest/pull/817

Visuals: https://github.com/yogstation13/Yogstation/pull/9195

TEG in this file uses different power generation, which requires some different setups for it to generate power. You cant just plug couple of thermomachines, set them to max with easy to make hypernoblium to make 200 MW of power, you can stack TEGS so it would even more power

Im using logistic function from Shiptest which will only start making power when temperatures are going to reach quite high, so you will need proper burn chamber or at least burn can. Its something different than pluging teg to few thermomachines and making more than SM multiplied by 100. 

/// A function that exponentially approaches a maximum value of L
/// k is the rate at which it approaches L, x_0 is the point where the function = 0
#define LOGISTIC_FUNCTION(L,k,x,x_0) (L/(1+(NUM_E**(-k*(x-x_0)))))

## Why It's Good For The Game

When SM explodes it will make most of the round run with no power or not enough to support whole station. Sometimes new engineers dont want to mess with SM because they are affraind of blowing it up. Thermo-electric generator was always a nice addon for extra power, Im not sure about the balance tho...might need to have efficiency tweaked a bit down...but its a one line of code changed...so far it can make 2MW with the setup in the bottom image.

Perhaps another engine for our modular engine PR, or just another project for our nice atmos engineers. TEG is tricky to set up, due to new pump limits when working with hight temperatures, but it can be worked with with special setups. 

## Changelog

:cl:
add: Added back TEG to techwebs and engineering protolathe
image: Changed TEG sprites from YogStation. Credits to original author.
/:cl:

### CODE REVIEW

**Comment by Erol509** [comment]
image

**Comment by ReturnToZender** [comment]
This looks SO cool. Can't wait to see it!

**Comment by Erol509** [comment]
> This looks SO cool. Can't wait to see it! You need to tick code\modules\power\generator.dm, though

Yeah, I will need to move the dmi file, etc... But my only concern is that overlay on the generator just dosnt want to work... Hmm, maybe power output is too low to even show it... Not sure. I might need to comment whole generator file, since it will cause problems when left together

File: code/modules/atmospherics/machinery/components/binary_devices/circulator.dm [hunk 1]
```diff
@@ -1,194 +1,198 @@
 //node2, air2, network2 correspond to input
 //node1, air1, network1 correspond to output
 
-/obj/machinery/atmospherics/components/binary/circulator
-	name = "circulator/heat exchanger"
-	desc = "A gas circulator pump and heat exchanger."
-	icon_state = "circ-off-0"
+// <--BUBBER removal START
```
[the remark this answers is not in the document]
Comment (Erol509) [reply code/modules/atmospherics/machinery/components/binary_devices/circulator.dm:4]: Alright, I didnt actually know that you can block comment. Will change that tomorrow

**Review by Erol509 — COMMENTED** [review]
Adds changes from the rewiew, still not all of them.. Will do it later

**Comment by Erol509** [comment]
Upstream has better version of this, since we are outdated as hell.

**Comment by Majkl-J** [comment]
image

Here is the current function for those curious, x is temperature difference [10kK], y is output [1MW]

**Review by Majkl-J — APPROVED** [review]

**Comment by Majkl-J** [comment]
> The biggest issue with the TEG before was people would simply plug 2 thermomachines into it and just win at power with no down side ever. I don't think the power or efficency should take a hit tho just make it start making power at 600 or 700 temp instead

It is still basically linear up until 5MW, and you only need a 3000K temp difference for like 1MW

**Pull request merged**